Openfam2.0

faizan-barmawer opened the pull request:
OpenFAM 2.0 documentation update:

     - Updated all man-page files for exception and c++ convention changes
     - Added man pages for fam_barrier_all and fam_stat APIs
     - Added pages for services and configuration files description
     - Updated index page with OpenFAM release information
     - Modification to OpenFAM 1.0 documentation include:
       - Added description and usage for fam_global descriptor in create_region
         and allocate data item man pages.
       - Added a note on usage of fam_quiet for all non-fetch atomic api man pages.
       - Added man-page for fam_size API
       - Corrected few spelling mistakes.

These changes are published from my private repository and can be viewed from this link:

https://faizan-barmawer.github.io/OpenFAM.github.io/

Once we accept these changes, we will copy all the documentation pages from parent directory to "v2.0" directory.

faizanbarmawer committed:
OpenFAM 2.0 documentation updates #1

OpenFAM 1.0 documentation saved in folder v1.0

faizanbarmawer committed:
OpenFAM 2.0 documentation updates #2

 - Updated all man-page files for exception and c++ convention changes
 - Added manpages for fam_barrier_all and fam_stat APIs
 - Added pages for services and configuration files description
 - Updated index page with OpenFAM release information
 - Modification to OpenFAM 1.0 documentation include:
   - Added description and usage for fam_global descriptor in create_region
     and allocate data item man pages.
   - Added a note on usage of fam_quiet for all non-fetch atomic api man pages.
   - Added man-page for fam_size API
   - Corrected few spelling mistakes.
 - Resolved all the review comments.

mashoodk commented on errors.html:
```diff
@@ -98,16 +102,16 @@ <h1>Error handling</h1>
-				types of failures, like <code>Runtime</code>, <code>Allocator</code>, <code>Data-path</code>, etc. It also
-				defines a list of individual error numbers to identify a specific
-				error. The exception object received by the caller in case of an
+			<p>The OpenFAM 2.0 implementation define a fam exception class, which
```
```suggestion
			<p>The OpenFAM 2.0 implementation defines a fam exception class, which
```

clarete-riana replied:
Fam_Exception  class ( for consistency)

mashoodk commented on errors.html:
```diff
@@ -232,30 +257,30 @@ <h1>Error handling</h1>
-      // Fam_Timeout_Exception, Fam_Datapath_Exception
+    	// This exception may actually result from a previous 
+    	// fam_put_nonblocking() operation with following error number
+        // FAM_ERR_TIMEOUT, FAM_ERR_OUTOFRANGE, FAM_ERR_NOPERM
```
extra spacing ?

mashoodk commented on fam_copy.html:
```diff
@@ -82,19 +85,20 @@ <h1>OpenFAM Reference Implementation</h1>
 		<article>
 			<h1 class="code">openfam::fam::fam_copy</h1>
-	<p>Create a second copy of a data item in FAM.</p>
+	<p>Copy data item from one FAM-resident data item to another FAM-resident data item within a different or same region.</p>
```
within a same region or across different regions

mashoodk commented on fam_stat.html:
```diff
@@ -0,0 +1,182 @@
+		Fam_Descriptor *arrayDescriptor = myFam->fam_lookup("myArray", "myOtherRegion");
+		
+		// Get data item info
+		fam_stat(arrayDescriptor, info);
```
missing , myFam->  ?

ssinghal53 commented:
Please separate release notes into a separate file from index.html

ssinghal53 commented on fam_barrier_all.html:
```diff
@@ -0,0 +1,156 @@
+	<p>The OpenFAM library suspends the execution of calling PE until all other
+		cooperating PEs issue a call to this function. Since data consistency in FAM cannot be
+		guaranteed by this library function, this method should not be used to synchronize the data for non-blocking APIs.</p>
+	<h2>Example</h2>
```
Unclear what this means. I believe you're saying that fam_barrier_all() should be preceded by a fam_quiet() to ensure that pending operations are completed before the barrier is reached.

ssinghal53 commented on services.html:
```diff
@@ -0,0 +1,153 @@
+at the network interfaces, as well as performance bottlenecks due to limited compute capacity at a given memory server.</li>
+<li>We need to consider dynamic addition of memory servers when additional memory is required, without
+disturbing previously allocated memory.</li>
+</ul>
```
Rephrase the above. Poor language style.

ssinghal53 commented on services.html:
```diff
@@ -0,0 +1,153 @@
+<ol>
+<li>All services running separately:
+If all the components are running as separate services, a grpc call will be made to the services.
+As shown in Figure 1, PE will make an RPC call to CIS, in-turn CIS will make an RPC call to metadata
```
Figure 1 does not show any RPC calls being made. Again No future tense for completed work.

ssinghal53 commented on services.html:
```diff
@@ -0,0 +1,153 @@
+</li>
+<li>CIS running as separate service with other services contained within CIS.
+In this case PE will make a grpc call to CIS and CIS makes direct  function call to
+other services. Figure 2, shows one such configuration where CIS and metadata running as a single service,
```
Same as before. Fig. 2 does not show any RPC calls. Please clean out all language in this file.

ssinghal53 commented on fam_initialize.html:
```diff
@@ -219,7 +226,7 @@ <h2 id="e1">Example</h2>
 	Fam_Options *fm = (Fam_Options*)malloc(sizeof(Fam_Options));
 	memset((void *)fm, 0, sizeof(Fam_Options));
```
What does the user need to do to pick only some options to be default? The description above seems to be an "all or nothing" behavior.

ssinghal53 commented on fam_get_option.html:
```diff
@@ -104,7 +108,7 @@ <h2>Input Arguments</h2>
 		</tbody>
 	</table>
-	The following optionNames are available:
+	The following option names are available:
```
VERSION below is returning "0.0.1". This default is incorrect-- should match the version of the OpenFAM library

ssinghal53 commented on fam_stat.html:
```diff
@@ -0,0 +1,182 @@
+    uint64_t size;
+    mode_t perm;
+    char *name;
+} Fam_Stat;
```
Add comments or text to describe the fields in the FAM_Stat structure. What is Fam_Stat.key?

ssinghal53 committed:
initial documentation updates

clarete-riana commented on index.html:
```diff
@@ -82,12 +85,47 @@ <h1>OpenFAM Reference Implementation</h1>
 		</nav>
 		<article>
+		<h1>OpenFAM release notes</h1>
+		<p>We are pleased to announce the release of OpenFAM 2.0 reference implementation. The highlights of the release is summarized below.
```
The highlights of the release are*

clarete-riana commented on index.html:
```diff
@@ -82,12 +85,47 @@ <h1>OpenFAM Reference Implementation</h1>
 		<article>
+		<h1>OpenFAM release notes</h1>
+		<p>We are pleased to announce the release of OpenFAM 2.0 reference implementation. The highlights of the release is summarized below.
+		Please refer NEWS.md for list of features, enhancements and bug fixes that are added to the source tree in all the released versions.
```
You may have to specify where NEWS.md can be found.

clarete-riana replied:
Also these two lines can be added as notes after the feature description for better continuity in the description.

clarete-riana commented on index.html:
```diff
@@ -82,12 +85,47 @@ <h1>OpenFAM Reference Implementation</h1>
+		Build and installation instructions are listed in README.md file located in the source tree.     </p>
+		OpenFAM version 2.0
+		<ul>
+			<li>Scalability improvements: Support for larger pool of memory by scaling to multiple memory servers. See <a href="services.html">services</a> for more details. </li>
```
Multiple memory server support and hence support for larger pools of memory spanning across memory servers.

clarete-riana commented on index.html:
```diff
@@ -82,12 +85,47 @@ <h1>OpenFAM Reference Implementation</h1>
+			<li>Scalability improvements: Support for larger pool of memory by scaling to multiple memory servers. See <a href="services.html">services</a> for more details. </li>
+			<li>Performance improvement of memory management and metadata operations: 
+				<ul>
+					<li>NVMM background worker thread is disabled from memory service.</li>
```
Looks like a NVMM change... Does it make sense to describe here?

clarete-riana commented on index.html:
```diff
@@ -82,12 +85,47 @@ <h1>OpenFAM Reference Implementation</h1>
+			<li>Performance improvement of memory management and metadata operations: 
+				<ul>
+					<li>NVMM background worker thread is disabled from memory service.</li>
+					<li>Delayed Free: Zeroing of memory have been moved from allocation path to free path by using delayed free thread.</li>  
```
Zeroing of memory has* been moved...

clarete-riana replied:
You may need to bring out how this will help... example say Zeroing of memory has been moved from performance sensitive allocation path to free path....

clarete-riana commented on index.html:
```diff
@@ -82,12 +85,47 @@ <h1>OpenFAM Reference Implementation</h1>
+			<li>OpenFAM APIs changes: 
+				<ul>
+					<li>fam_size API is replaced with fam_stat API.</li>
+					<li>Modified fam_copy API to copy data between different regions.</li>
```
Is this enhancement to the existing capability of fam_copy or modification to its existing behavior? I think it is the first, and it needs to come out from the description.

clarete-riana commented on errors.html:
```diff
@@ -98,16 +102,16 @@ <h1>Error handling</h1>
+				error condition. The fam exception object received by the caller in case of an
 				error will contain a specific error number and an appropriate error
 				message string. The application can retrieve this information using
 				member functions, <code>fam_error()</code> and <code>fam_error_msg()/what()</code> and take
```
there is no need for , here after member functions.

clarete-riana commented on errors.html:
```diff
@@ -98,16 +102,16 @@ <h1>Error handling</h1>
-				error. The exception object received by the caller in case of an
+			<p>The OpenFAM 2.0 implementation define a fam exception class, which
+				is derived from C++ standard exception class. It also defines a list of individual error
+				numbers to categorize various types of failures, like <code>Runtime</code>, <code>Allocator</code>,
```
No need for , after failures

clarete-riana commented on errors.html:
```diff
@@ -98,16 +102,16 @@ <h1>Error handling</h1>
+				is derived from C++ standard exception class. It also defines a list of individual error
+				numbers to categorize various types of failures, like <code>Runtime</code>, <code>Allocator</code>,
+				<code>Data-path</code>, etc. Individual error number identifies a specific
+				error condition. The fam exception object received by the caller in case of an
```
The fam exception object received by the caller, in case of an error, will contain...

clarete-riana commented on errors.html:
```diff
@@ -214,15 +203,51 @@ <h1>Error handling</h1>
-				the following exceptions - <code>Fam_InvalidOption_Exception</code>,
-				<code>Fam_Allocator_Exception</code>, <code>Fam_Datapath_Exception</code> or
-				<code>Fam_Timeout_Exception</code>. In general, the application should use the
+				<code>fam_put_blocking()</code> will either complete successfully, or throw Fam_Exception object 
```
Fam_Exception object is returned or thrown?

clarete-riana commented on services.html:
```diff
@@ -0,0 +1,153 @@
+		<article>
+		<h1>OpenFAM 2.0 Services</h1>
+<p> The initial reference implementation of OpenFAM was built using a single memory server.
+ As expected, the memory server becomes a bottleneck even for a small number of concurrent PEs.
```
The memory server soon becomes a bottleneck with the increase in number of concurrent PEs.

clarete-riana commented on services.html:
```diff
@@ -0,0 +1,153 @@
+the CIS will communicate with memory servers via an RPC service or via direct functions calls.</li>
+</ol>
+We have defined an RPC path and a direct path for all the three components – CIS, memory management service and metadata management service.
+Yaml based configuration parameters are available to decide how these services are running.  If the components are running as separate services,
```
available to define* how these servcies..

clarete-riana commented on services.html:
```diff
@@ -0,0 +1,153 @@
+For memory management and metadata operations, like allocate, lookup, create_region, etc.,
+the CIS will communicate with memory servers via an RPC service or via direct functions calls.</li>
+</ol>
+We have defined an RPC path and a direct path for all the three components – CIS, memory management service and metadata management service.
```
We have defined a* RPC path and a direct path for interactions between the three components...

clarete-riana commented on services.html:
```diff
@@ -0,0 +1,153 @@
+</ol>
+We have defined an RPC path and a direct path for all the three components – CIS, memory management service and metadata management service.
+Yaml based configuration parameters are available to decide how these services are running.  If the components are running as separate services,
+a grpc call is made to the service. If the services are running within other component (direct path), a function call is made.
```
If components/services are collocated within the same node, they interact through direct function calls.

clarete-riana commented on services.html:
```diff
@@ -0,0 +1,153 @@
+<p> Now let us look at some of the possible combinations.
+<ol>
+<li>All services running separately:
+If all the components are running as separate services, a grpc call will be made to the services.
```
Throughout this description the concept of separate service is confusing to me. so even if the services are running on the same node we are making GPRC calls? That can be avoided right? Unless there are other reasons fro doing this. We need to think through this design a bit more...

clarete-riana commented on services.html:
```diff
@@ -0,0 +1,153 @@
+We have defined an RPC path and a direct path for all the three components – CIS, memory management service and metadata management service.
+Yaml based configuration parameters are available to decide how these services are running.  If the components are running as separate services,
+a grpc call is made to the service. If the services are running within other component (direct path), a function call is made.
+<p> Now let us look at some of the possible combinations.
```
Below are some of the possible combinations:

clarete-riana commented on services.html:
```diff
@@ -0,0 +1,153 @@
+management service and memory management service for all memory management and metadata operations.
+</li>
+<li>CIS running as separate service with other services contained within CIS.
+In this case PE will make a grpc call to CIS and CIS makes direct  function call to
```
function calls*

clarete-riana commented on services.html:
```diff
@@ -0,0 +1,153 @@
+<img src="images/openfam2.0exampleconfiguration.png" width="450px">
+<figcaption>Figure 2: A possible configuration with metadata running as part of CIS</figcaption>
+</figure>
+<p>See configuration and startup of OpenFAM 2.0 services in this section. <a href="config_files.html">Configuration Files</a></p>
```
The ". " should be after Configuration Files.

ssinghal53 committed:
documentation updates

ssinghal53 committed:
more updates

mashoodk commented on fam_allocate.html:
```diff
@@ -110,18 +115,24 @@ <h2>Input Arguments</h2>
 					Descriptor for the region where the data item is being
 						created. If null, the default specified during fam_initialize() is
-						used.</td></tr>
+						used.<span style="color:red;font-weight:bold">Is this true? <code>fam_initialize</code> says that default region name is not used. Do we still have a default region?</span></td></tr>
```
@faizan-barmawer we do not have default region name. We can remove this sentence.

faizanbarmawer committed:
minor updates and address comments from Sharad.

ssinghal53 commented on fam_create_region.html:
```diff
@@ -142,20 +147,39 @@ <h2>Input Arguments</h2>
-		limited to 16 EiB by the API.</p>
+		limited to 16 EiB by the API. See <a href="limitations.html">Design Choices</a> for the current design limits within the OpenFAM reference implementation.
+		</p>
+	<p>Fam_Region_Descriptor provides a method to retrieve an opaque read-only data structure called 
```
Change "Fam_Region_Descriptor" on line 170 to <code>Fam_Region_Descriptor</code>.

ssinghal53 commented on fam_create_region.html:
```diff
@@ -142,20 +147,39 @@ <h2>Input Arguments</h2>
+// Fam_Global_Descriptor Fam_Region_Descriptor::get_global_descriptor();
+Fam_Global_Descriptor gDescriptor = regionDescriptor->get_global_descriptor();
+</pre>
+	<p>Fam_Region_Descriptor can be instantiated from Fam_Global_Descriptor using the following constructor.</p>
```
Line 178, insert code /code blocks around Fam_Region_Descriptor/Fam_Global_Descriptor for consistent formatting.

ssinghal53 commented on fam_allocate.html:
```diff
@@ -133,6 +143,20 @@ <h2>Notes</h2>
 		provided during allocation, then the program itself is responsible for
 		saving and tracking the descriptor if required later, since the data
 		in FAM will be unreachable without the descriptor.</p>
+	<p><code>Fam_Descriptor</code> provides a method to retrieve an opaque read-only data structure called 
```
Line 142 (Notes) add code /code block around fam_lookup()

ssinghal53 commented on fam_change_permissions.html:
```diff
@@ -161,7 +169,7 @@ <h2>Example</h2>
 		</article>
 	</section>
 	<footer>
-		<p>Copyright 2019, Hewlett Packard Enterprise Development Co, LLP</p>
```
Line 132, Notes should say that permissions are checked on first descriptor use in a running program. Changed permissions are only applied the next time the program is run, not immediately on next access.

ssinghal53 commented on fam_change_permissions.html:
```diff
@@ -161,7 +169,7 @@ <h2>Example</h2>
 	</section>
 	<footer>
-		<p>Copyright 2019, Hewlett Packard Enterprise Development Co, LLP</p>
+		<p>Copyright 2021, Hewlett Packard Enterprise Development Co, LLP</p>
```
Delete three lines starting at 159 (// free allocated space in FAM) in this example. If the fam is deallocated, the change in permissions is meaningless for the next program.

faizanbarmawer committed:
Addressed the following review comments:
 - Added <code></code> block around API/descriptor names to have consistent formatting style.
 - As suggested, added a note in fam_change_permission manpage.

faizan-barmawer commented:
Hi Sharad,

I have addressed your recent comments and updated this PR.

    - Added <code></code> block around API/descriptor names to have consistent formatting style.
    - As suggested, added a note in fam_change_permission man page.

Regards,
Faizan

faizanbarmawer committed:
Addressed comments to fix example code issues.
